### ISSUE TITLE
Add a tool to update ninja's information after generation

### ISSUE THREAD

**Issue by mathstuf** [ninja-build/ninja#906]
When Ninja files are generated by a tool for the first time, if it has `restat = 1` set on the `generator = 1` rule, Ninja will rerun the generate tool needlessly solely to update its deps database. It would be nice if the generator tool could tell Ninja to update its information right after generation so that it doesn't need to run again.

**Comment by mathstuf** [comment]
Something like `ninja -t initdb`?

**Comment by nico** [comment]
Do you have a smal example manifest file where this happens?

**Comment by mathstuf** [comment]
So the problem is that a build tool which saves a cache for configuration values for the user which writes at the end of "configure" (as expected) lists the cache file as an input to the generator and any configured files (`foo.ext.in` → `foo.ext`) files as outputs to the generate step. The fact that the (input) cache is newer than the (output) configured file triggers ninja to do the generation of the build itself and from then on the `restat` mechanism prevents further reruns of the tool. The idea is for the tool to be able to tell ninja that all is well with the build and to look at the generate step and do its `restat` logic there.

Example files:

`input.txt.in`:
```
REPLACE_ME
```

`build.ninja.in`:
```ninja
# REPLACE_ME

build all: phony

default all

rule regenerate
    command = /bin/sh mkninja.sh
    description = Generating the ninja file...
    restat = 1
    generator = 1

build build.ninja input.txt: regenerate | input.txt.in mkninja.sh build.ninja.in build.conf
```

`mkninja.sh`:
```sh
#!/bin/sh

echo "Regenerating..."

# Read in the configuration
if [ -e build.conf ]; then
    txt="$( cat build.conf )"
else
    txt="default"
fi

# Conservatively write out a file. If this is not done, the configuration file
# is *always* written to and since it is an input to the generate step, it will
# *always* run.
conservative_replace () {
    fout="$1"
    shift

    fin="$fout.tmp"

    if ! [ -e "$fout" ]; then
        # Use the input if the output doesn't exist.
        mv "$fin" "$fout"
    elif ! diff -q "$fin" "$fout"; then
        # Use the input if they differ.
        mv "$fin" "$fout"
    else
        # No updates; remove the input file.
        rm "$fin"
    fi
}

# Configure a file.
sed -e 's/REPLACE_ME/'"$txt"'/g' < input.txt.in > input.txt.tmp
conservative_replace input.txt

sleep 2 # simulate lots of work.

# Write configuration for future use.
echo "$txt" > build.conf.tmp
conservative_replace build.conf

# Write the build file.
sed -e 's/REPLACE_ME/'"$txt"'/g' < build.ninja.in > build.ninja.tmp
conservative_replace build.ninja
```

**Comment by maximuska** [comment]
Its not clear why manifest depends on build.conf which is updated when the
manifest is generated, which is the fundamental problem, I think.. No?

-- brevity due to the phone

**Comment by mathstuf** [comment]
If `build.conf` is updated, the tool needs to be rerun, so it /needs/ to be an input. I don't think any other build rules depending on `build.conf` makes any sense because they should just get the information necessary during the configure step anyways, so listing it as an output isn't necessary or sensible (and it would only make circular deps).

**Comment by Boddlnagg** [comment]
The way that I deal with this in my custom generator is that when the generator runs for the first time, it generates only a "stub" ninja buildfile, which contains no rules except the one to run the generator (I call this process "bootstrapping"). Then, on the first invocation of the `ninja` command, the generator will actually run, and the deps database is correctly updated.

**Comment by NiklasRosenstein** [comment]
@Boddlnagg Thanks for that tip. I've now worked around it using a similar approach. Unfortunately it requires quite a number of hacks.

When freshly generating the Ninja manifest, the generator command is special and checks if any of the dependencies have actually changed. If they are not, the step is skipped (which is the case when Ninja invokes the generator for the first time because it has no entry in `.ninja_deps`).

When Ninja actually invokes the generator when the inputs are changed, the special generator command will notice this and invoke the real generator command. This time, a flag is set so that the real generator command is also written into the Ninja file instead of the special command.

I use `.bat` files on Windows to write the commands (as the command may have additional environment variables or a different CWD), even for the generator command. One problem I encountered was that the generator command rewrote the batch file that invoked the generator. This lead to unexpected odd behaviour as the Batch file appaers to be read as it is executed.

Ping-ponging this batch file for the generator (ie. switching between two alternative filenames every generator invokation) solves this issue.
